File: packages/shared/py/solarreach_shared/financial.py
"""Financial / scoring math for SolarReach.

All sums in GBP; energy in kWh; rates per kWh.
"""

from __future__ import annotations

from . import constants


def _clamp(x: float, lo: float = 0.0, hi: float = 1.0) -> float:
    return max(lo, min(hi, x))


def composite_score(solar_roi: float, fin_health: float, social_impact: float) -> int:
    """Weighted composite score in 0..100.

    Inputs are unit-interval [0,1]; out-of-range values are clamped before weighting.
    Weights: solar_roi 0.5, financial_health 0.3, social_impact 0.2.
    """
    s = (
        _clamp(solar_roi) * constants.SCORE_WEIGHT_SOLAR_ROI
        + _clamp(fin_health) * constants.SCORE_WEIGHT_FIN_HEALTH
        + _clamp(social_impact) * constants.SCORE_WEIGHT_SOCIAL_IMPACT
    )
    return int(round(s * 100))


def roi_gate(score: int, threshold: int = constants.SCORE_THRESHOLD_DEFAULT) -> bool:
    """Pass/fail gate over a composite score (>= threshold)."""
    return score >= threshold


def capex(panel_count: int) -> float:
    """Estimated capex (GBP) including 4% installer margin.

    capex = (panel_count * panel_unit_cost + rated_kw * install_per_kw) * (1 + margin)
    """
    if panel_count <= 0:
        return 0.0
    rated_kw = panel_count * constants.PANEL_RATED_KW
    base = (
        panel_count * constants.PANEL_UNIT_COST_GBP
        + rated_kw * constants.INSTALL_COST_PER_KW_GBP
    )
    return base * (1.0 + constants.INSTALLER_MARGIN_PCT)


def annual_saving_gbp(annual_kwh: float) -> float:
    """Modelled annual saving = self-consumed kWh * import rate
    + exported kWh * SEG export rate.
    """
    if annual_kwh <= 0:
        return 0.0
    self_consumed = annual_kwh * constants.SELF_CONSUME_FRACTION
    exported = annual_kwh - self_consumed
    return (
        self_consumed * constants.UK_IMPORT_RATE_GBP_PER_KWH
        + exported * constants.UK_SEG_EXPORT_RATE_GBP_PER_KWH
    )


def payback_years(capex_gbp: float, annual_saving_gbp: float) -> float:
    """Simple payback. Returns +inf if saving is non-positive."""
    if annual_saving_gbp <= 0:
        return float("inf")
    return capex_gbp / annual_saving_gbp


def npv_25yr(capex_gbp: float, annual_saving_gbp: float) -> float:
    """25-year NPV with 0.5%/yr panel degradation, fixed discount rate.

    NPV = -capex + sum_{t=1..25} saving_t / (1+r)^t
    saving_t = annual_saving_gbp * (1 - degradation)^(t-1)
    """
    npv = -float(capex_gbp)
    for t in range(1, constants.NPV_HORIZON_YEARS + 1):
        deg = (1.0 - constants.PANEL_DEGRADATION_PER_YEAR) ** (t - 1)
        cf = annual_saving_gbp * deg
        disc = (1.0 + constants.DISCOUNT_RATE) ** t
        npv += cf / disc
    return npv


if __name__ == "__main__":
    # Smoke run
    print("composite_score(0.8, 0.6, 0.4) =", composite_score(0.8, 0.6, 0.4))
    print("roi_gate(74) =", roi_gate(74))
    print("capex(24 panels) = £", round(capex(24), 2))
    print("annual_saving(10000 kWh) = £", round(annual_saving_gbp(10000), 2))
    print(
        "payback_years(20000, 2500) =",
        round(payback_years(20000.0, 2500.0), 2),
    )
    print(
        "npv_25yr(20000, 2500) = £",
        round(npv_25yr(20000.0, 2500.0), 2),
    )

File: packages/shared/ts/src/financial.ts
// JS mirror of solarreach_shared/financial.py — parity within ±£1.

export const SCORE_WEIGHT_SOLAR_ROI = 0.5;
export const SCORE_WEIGHT_FIN_HEALTH = 0.3;
export const SCORE_WEIGHT_SOCIAL_IMPACT = 0.2;
export const SCORE_THRESHOLD_DEFAULT = 70;

export const PANEL_UNIT_COST_GBP = 850.0;
export const INSTALL_COST_PER_KW_GBP = 180.0;
export const PANEL_RATED_KW = 0.42;
export const INSTALLER_MARGIN_PCT = 0.04;

export const UK_SEG_EXPORT_RATE_GBP_PER_KWH = 0.15;
export const UK_IMPORT_RATE_GBP_PER_KWH = 0.27;
export const SELF_CONSUME_FRACTION = 0.5;
export const DISCOUNT_RATE = 0.05;
export const NPV_HORIZON_YEARS = 25;
export const PANEL_DEGRADATION_PER_YEAR = 0.005;

const clamp = (x: number, lo = 0, hi = 1): number =>
  Math.max(lo, Math.min(hi, x));

export function compositeScore(
  solarRoi: number,
  finHealth: number,
  socialImpact: number,
): number {
  const s =
    clamp(solarRoi) * SCORE_WEIGHT_SOLAR_ROI +
    clamp(finHealth) * SCORE_WEIGHT_FIN_HEALTH +
    clamp(socialImpact) * SCORE_WEIGHT_SOCIAL_IMPACT;
  return Math.round(s * 100);
}

export function roiGate(
  score: number,
  threshold: number = SCORE_THRESHOLD_DEFAULT,
): boolean {
  return score >= threshold;
}

export function capex(panelCount: number): number {
  if (panelCount <= 0) return 0;
  const ratedKw = panelCount * PANEL_RATED_KW;
  const base =
    panelCount * PANEL_UNIT_COST_GBP + ratedKw * INSTALL_COST_PER_KW_GBP;
  return base * (1 + INSTALLER_MARGIN_PCT);
}

export function annualSavingGbp(annualKwh: number): number {
  if (annualKwh <= 0) return 0;
  const selfConsumed = annualKwh * SELF_CONSUME_FRACTION;
  const exported = annualKwh - selfConsumed;
  return (
    selfConsumed * UK_IMPORT_RATE_GBP_PER_KWH +
    exported * UK_SEG_EXPORT_RATE_GBP_PER_KWH
  );
}

export function paybackYears(
  capexGbp: number,
  annualSavingGbp: number,
): number {
  if (annualSavingGbp <= 0) return Infinity;
  return capexGbp / annualSavingGbp;
}

export function npv25yr(
  capexGbp: number,
  annualSavingGbp: number,
): number {
  let npv = -capexGbp;
  for (let t = 1; t <= NPV_HORIZON_YEARS; t++) {
    const deg = (1 - PANEL_DEGRADATION_PER_YEAR) ** (t - 1);
    const cf = annualSavingGbp * deg;
    const disc = (1 + DISCOUNT_RATE) ** t;
    npv += cf / disc;
  }
  return npv;
}

// CLI smoke
if (
  typeof process !== "undefined" &&
  process.argv?.[1]?.endsWith("financial.ts")
) {
  console.log("compositeScore(0.8, 0.6, 0.4) =", compositeScore(0.8, 0.6, 0.4));
  console.log("capex(24) =", capex(24).toFixed(2));
  console.log("npv25yr(20000, 2500) =", npv25yr(20000, 2500).toFixed(2));
}
